test/jdk/javax/swing/border/EtchedBorder/ScaledEtchedBorderTest.java

```diff
@@ -66,165 +71,256 @@ public class ScaledEtchedBorderTest {
+                            state = State.INSIDE;
+                        } else {
+                            throwUnexpectedColor(x, y, color);
+                        }
+                        break;
+
+                    case INSIDE:
+                        if (color == SHADOW.getRGB()) {
+                            state = State.RIGHT_SHADOW;
+                            borderThickness = 1;
+                        } else if (color != INSIDE_COLOR.getRGB()) {
+                            throwUnexpectedColor(x, y, color);
+                        }
+                        break;
+
+                    case RIGHT_SHADOW:
+                        if (color == SHADOW.getRGB()) {
+                            borderThickness++;
+                        } else if (color == HIGHLIGHT.getRGB()) {
+                            if (borderThickness != thickness) {
+                                throwWrongThickness(thickness, borderThickness, x, y);
+                            }
+                            borderThickness = 1;
+                            state = State.RIGHT_HIGHLIGHT;
+                        } else {
+                            throwUnexpectedColor(x, y, color);
+                        }
+                        break;
+
+                    case RIGHT_HIGHLIGHT:
+                        if (color == HIGHLIGHT.getRGB()) {
+                            borderThickness++;
+                        } else if (color == OUTER_COLOR.getRGB()) {
+                            if (borderThickness != thickness) {
+                                throwWrongThickness(thickness, borderThickness, x, y);
+                            }
+                            borderThickness = 0;
+                            state = State.BACKGROUND;
+                        } else {
+                            throwUnexpectedColor(x, y, color);
+                        }
+                        break;
                 }
-            }
-        }
+            } while (yStep > 0 && ((y += yStep) < height));
+        } while (xStep > 0 && ((x += xStep) < width));
+    }
+
+    private enum State {
+        BACKGROUND,
+        LEFT_SHADOW, LEFT_HIGHLIGHT,
+        INSIDE,
+        RIGHT_SHADOW, RIGHT_HIGHLIGHT
+    }
+
+    private static void throwWrongThickness(int thickness, int borderThickness,
+                                            int x, int y) {
+        throw new Error(
+                String.format("Wrong border thickness at %d, %d: %d vs %d",
+                              x, y, borderThickness, thickness));
     }
 
-    private static void createGUI(boolean show) {
-        // Render content panel
-        JPanel contentPanel = new JPanel();
-        contentPanel.setLayout(new BoxLayout(contentPanel, BoxLayout.Y_AXIS));
+    private static void throwUnexpectedColor(int x, int y, int color) {
+        throw new Error(
+                String.format("Unexpected color at %d, %d: %08x",
+                              x, y, color));
```

Comment:
Same as above - including state value in error msg can be helpful here.

Reply:
I don't think it's needed: the stack trace of the exception contains the line number which uniquely identifies the state.

For example, an excerpt of test output when it fails:
```text
Scaling: 1.00
java.lang.Error: Unexpected color at 62, 24: ffffffff
        at ScaledEtchedBorderTest.throwUnexpectedColor(ScaledEtchedBorderTest.java:255)
        at ScaledEtchedBorderTest.checkBorder(ScaledEtchedBorderTest.java:231)
        at ScaledEtchedBorderTest.checkVerticalBorders(ScaledEtchedBorderTest.java:131)
        at ScaledEtchedBorderTest.verifyBorderRendering(ScaledEtchedBorderTest.java:99)
        at ScaledEtchedBorderTest.testScaling(ScaledEtchedBorderTest.java:87)
        at ScaledEtchedBorderTest.lambda$main$0(ScaledEtchedBorderTest.java:77)
	at …
```

https://github.com/openjdk/jdk/blob/b78389e8c1e23cb83338387ba20091bf705480e1/test/jdk/javax/swing/border/EtchedBorder/ScaledEtchedBorderTest.java#L221-L232

Line 231 corresponds to the `RIGHT_HIGHLIGHT` state. If you open the `test1.00.png` image, you quickly see, there's _a white line_ below the highlight line which shouldn't be there.

Therefore, passing the additional parameter doesn't add much value.